Operator/CES: Simplify rate limit configuration

thorn3r opened the pull request:
The CiliumEndpointSlice controller has 6 rate limit related config options. This PR attempts to simplify configuration by reducing that set to 2: one for supplying a custom, dynamic rate limit config, and another to supply a base multiplier for simple tuning of an existing config.

Note: The supplied default dynamic rate limit may need some adjustment. This will likely be updated after future scale testing and/or PR feedback.

See individual commits for more details

```release-note
Simplify rate limit configuration options for the CiliumEndpointSlice controller.
```

tklauser reviewed (changes requested):
I haven't looked at the entire PR in detail but one thing that stood out on initial review is that the old options and Helm values are removed rather than deprecated.

I think our current policy is to deprecate config options/values in Cilium release v1.n and and introduce the new options/values in addition. The deprecation of the options/values is documented in the v1.n release upgrade section.

In Cilium release v1.n+1 we'd then remove the deprecated options/values and solely rely on the new configuration and again document the removal.

I understand this process makes this PR a bit more complicated to implement but will allow users to upgrade gradually to the new options rather than on upgrade breaking their existing configuration relying on the old options/values.

giorio94 commented on install/kubernetes/cilium/templates/validate.yaml:
```diff
@@ -73,9 +73,9 @@
 
 {{/* validate Cilium operator */}}
-{{- if eq .Values.enableCiliumEndpointSlice true }}
+{{- if eq .Values.ciliumEndpointSlice.enabled true }}
```
Should account for both the new and the deprecated setting.

giorio94 commented on operator/pkg/ciliumendpointslice/endpointslice.go:
```diff
@@ -153,11 +151,13 @@ func (c *Controller) runCiliumEndpointsUpdater(ctx context.Context) error {
 			c.logger.WithFields(logrus.Fields{
-				logfields.CEPName: event.Key.String()}).Debug("Got Upsert Endpoint event")
+				logfields.CEPName: event.Key.String(),
+			}).Debug("Got Upsert Endpoint event")
```
These changes seem unrelated, and should be reverted.

giorio94 commented on operator/pkg/ciliumendpointslice/rate_limit.go:
```diff
@@ -29,87 +30,68 @@ type rateLimitConfig struct {
+	if len(p.Cfg.CESDynamicRateLimitConfig) != 0 {
+		parsed, err := parseDynamicRateLimit(p.Cfg.CESDynamicRateLimitConfig)
 		if err != nil {
 			p.Logger.WithError(err).Warn("Couldn't parse dynamic rate limit config")
```
It's personally suggest to bail out loudly if the configuration cannot be parsed, to avoid surprises to users.

thorn3r replied:
Done :+1:

giorio94 commented on test/helpers/kubectl.go:
```diff
@@ -508,7 +508,6 @@ func (kub *Kubectl) ParallelResourceDelete(namespace, resource string, names []s
 			if !res.WasSuccessful() {
 				ginkgoext.By("Unable to delete %s %s with '%s': %s",
 					resource, name, cmd, res.OutputPrettyPrint())
-
```
Most of the changes in this file appear to be unrelated, and should be reverted.

thorn3r replied:
It seems I unknowingly had `gofumpt` enabled :sweat_smile:

giorio94 commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -707,8 +707,18 @@ enableCiliumEndpointSlice: false
   enabled: false
-  # -- Configure a custom dynamic rate limit for the CiliumEndpointSlice controller.
-  dynamicRateLimitConfig: ""
+  # -- List of rate limit options to be used for the CiliumEndpointSlice controller
```
I'd suggest to squash this commit with the first one, as strongly related.

giorio94 commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -707,8 +707,18 @@ enableCiliumEndpointSlice: false
+  # # -- The sustained request rate in requests per second
+  #   limit: 10
+  # # -- The burst request rate in requests per second
+  #   burst: 20
```
Looking at the code, it seems that the maximum allowed values are respectively 50 and 100, which may be worth documenting.

thorn3r replied:
Added in the comments. I was hoping helm-docs supported docstrings for fields of an object within a list, but I think that may beyond what the tool can handle. Added details for the fields within the top-level comment and moved the default config to the helm values..

giorio94 commented on operator/pkg/ciliumendpointslice/rate_limit.go:
```diff
@@ -29,87 +30,68 @@ type rateLimitConfig struct {
+	{Nodes: 0, Limit: 10, Burst: 20},
+	{Nodes: 100, Limit: 7, Burst: 15},
+	{Nodes: 500, Limit: 5, Burst: 10},
+}
```
I personally find a complex default configuration like this to be possibly confusing for users, as there's no easy way to figure it out except looking at the code. How about having a single limit/burst default, and then overriding it by default through the helm values? That way, the default values would be immediately visible to the users, and tuning would also be way simpler. That approach would also allow to drop the additional multiplier flag introduced in the second commit, as the values could be tuned directly if necessary.

thorn3r replied:
SGTM :+1: I think that's a more user-friendly approach :+1:

giorio94 commented on install/kubernetes/cilium/values.yaml.tmpl:
```diff
@@ -707,8 +707,18 @@ enableCiliumEndpointSlice: false
+  # A custom dynamic rate limit configuration can be provided for the
+  # CiliumEndpointSlice controller. Each object in the list must have a
+  # node count, limit, and burst value.
+  # dynamicRateLimitConfig:
```
I'd personally strip the term `dynamic` from the various settings/flags, considering that it's the only rate limit configuration available.

thorn3r replied:
Good point, updated to make this change

giorio94 commented on Documentation/operations/upgrade.rst:
```diff
@@ -340,6 +340,12 @@ Annotations:
+  configuration, replacing the following flags which have been deprecated:
+  ``ces-write-qps-limit``, ``ces-write-qps-burst``, ``ces-enable-dynamic-rate-limit``,
+  ``ces-dynamic-rate-limit-nodes``, ``ces-dynamic-rate-limit-qps-limit``,
+  ``ces-dynamic-rate-limit-qps-burst``
```
I'd be more explicit in the wording here, as the deprecated values seem to be effectively ignored with the current implementation, which is not what a user would likely expect reading *deprecated*. I personally don't have a strong opinion on preserving the existing functionality considering that CiliumEndpointSlices are a beta feature, but it should be stated clearly (I'd also ensure that the operator logs errors in case the deprecated flags are used if the values are ignored).

thorn3r replied:
Done for both, let me know what you think

giorio94 replied:
:+1: Thanks!

giorio94 reviewed (changes requested):
Nice work, always in favor of simplifying the Cilium's configuration!

I've left a few suggestions inline, mainly about the possibility of making the default settings more explicit, further removing the multiplier setting, and clarifying the semantic around the *deprecated* wording.

thorn3r committed:
operator/CES: simplify rate limit configuration for CES controller

Cilium Operator has 6 flags related to configuring the client work queue
rate-limiter for the CiliumEndpointSlice controller. This commit
deprecates the existing flags and adds a single new flag for supplying a
custom dynamic rate limit configuration. The following flags are now deprecated
and will be removed in a future release:

```
--ces-write-qps-limit
--ces-write-qps-burst
--ces-enable-dynamic-rate-limit
--ces-dynamic-rate-limit-nodes
--ces-dynamic-rate-limit-qps-limit
--ces-dynamic-rate-limit-qps-burst
```

The `enableCiliumEndpointSlice` field in the helm values
has been deprecated and replaced by `ciliumEndpointSlice.enabled`.
`enableCiliumEndpointSlice` will be removed in a future release.

The static rate limit options have been removed in favor of always using
a dynamic rate limit. The default rate limits defined in
`ciliumEndpointSlice.rateLimits` should be sufficient
for most clusters, however a custom configuration can be supplied for
specific use-cases.

The new flag 'ces-rate-limits' accepts a string containing
a JSON list of 'rateLimit' objects, each containing a limit and burst
setting, and the number of nodes at which the rate limit should be
applied. Example:

"[{\"nodes\": 5, \"limit\": 15.0, \"burst\": 30}]"

Custom rate limits can also be supplied via the Helm
value `ciliumEndpointSlice.rateLimits`, which accepts a yaml
list of rate limit objects that will be formatted into a properly
escaped JSON string.

Signed-off-by: Tim Horner <timothy.horner@isovalent.com>

tklauser reviewed (approved):
Thanks @thorn3r!

giorio94 commented on operator/pkg/ciliumendpointslice/rate_limit.go:
```diff
@@ -22,94 +23,73 @@ type rateLimit struct {
+		parsed, err := parseDynamicRateLimit(p.Cfg.CESDynamicRateLimitConfig)
 		if err != nil {
-			p.Logger.WithError(err).Warn("Couldn't parse dynamic rate limit config")
+			p.Logger.WithError(err).Fatal("Couldn't parse CES rate limit config")
```
Nit. AFAIR it is better to return an error to hive, rather than calling Fatal directly, to enable proper shutdown.

thorn3r replied:
Updated the surrounding functions to bubble the error up to hive

giorio94 commented on operator/pkg/ciliumendpointslice/rate_limit.go:
```diff
@@ -22,94 +23,73 @@ type rateLimit struct {
+func parseDynamicRateLimit(cfg string) (dynamicRateLimit, error) {
+	if len(cfg) == 0 {
+		return nil, nil
 	}
```
Nit. This check appears to be redundant.

giorio94 commented on operator/pkg/ciliumendpointslice/cell.go:
```diff
@@ -65,29 +70,34 @@ type Config struct {
-	CESDynamicRateLimitQPSBurst: []string{},
+	CESMaxCEPsInCES:           100,
+	CESSlicingMode:            "cesSliceModeIdentity",
+	CESDynamicRateLimitConfig: "",
```
NIt. It may be worth to display the actual default configuration, rather than an empty string.

thorn3r replied:
I removed the default rate limit struct and put the default values here instead.

giorio94 reviewed (approved):
lgtm, thanks! Just a couple of final nits inline.
